core/src/main/java/org/apache/calcite/plan/MaterializedViewSubstitutionVisitor.java

```diff
@@ -54,6 +55,15 @@ public MaterializedViewSubstitutionVisitor(RelNode target_, RelNode query_,
     super(target_, query_, EXTENDED_RULES, relBuilderFactory);
   }
 
+  /**
+   * A method used to add a user defined {@link SubstitutionVisitor.UnifyRule},
+   * thus to extend the ability of substitution based materialization
+   * matching in different scenarios.
+   */
+  public static void addMatchingRule(UnifyRule rule) {
+    EXTENDED_RULES.add(rule);
+  }
+
```

Comment:
Some suggestions:
1. How about add an instance method instead of static class method, and keep `EXTENDED_RULES` immutable as it is?
So that, different instances of MaterializedViewSubstitutionVisitor will not affect each other.
2. If we have `addMatchingRule` method, do we need to have a coresponding`removeMatchingRule` method?

Reply:
Thanks yanlin ~
I think your concern is correct.
I updated. In current change, the rules is registered from `RelOptPlanner` and passed through `RelOptMaterializations` to `MaterializedViewSubstitutionVisitor`.
Thus user will set up the self defined materialization rules when preparing `VolcanoPlanner`